x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/QueryBuilderResolver.java

```diff
@@ -0,0 +1,170 @@
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    private Set<FullTextFunction> fullTextFunctions(LogicalPlan plan) {
+        Set<FullTextFunction> functions = new HashSet<>();
+        plan.forEachExpressionDown(FullTextFunction.class, func -> functions.add(func));
+        return functions;
+    }
+
+    public Set<String> indexNames(LogicalPlan plan) {
+        Holder<Set<String>> indexNames = new Holder<>();
+
+        plan.forEachDown(EsRelation.class, esRelation -> { indexNames.set(esRelation.index().concreteIndices()); });
+
+        return indexNames.get();
+    }
+
+    public LogicalPlan planWithResolvedQueryBuilders(LogicalPlan plan, Map<FullTextFunction, QueryBuilder> newQueryBuilders) {
+        LogicalPlan newPlan = plan.transformExpressionsDown(FullTextFunction.class, m -> {
+            if (newQueryBuilders.keySet().contains(m)) {
+                return m.replaceQueryBuilder(newQueryBuilders.get(m));
+            }
+            return m;
+        });
+        // The given plan was already analyzed and optimized, so we set the resulted plan to optimized as well.
+        newPlan.setOptimized();
+        return newPlan;
+    }
+
+    private QueryRewriteContext queryRewriteContext(Set<String> indexNames) {
+        ResolvedIndices resolvedIndices = ResolvedIndices.resolveWithIndexNamesAndOptions(
+            indexNames.toArray(String[]::new),
+            IndexResolver.FIELD_CAPS_INDICES_OPTIONS,
+            clusterService.state(),
+            indexNameExpressionResolver,
+            transportService.getRemoteClusterService(),
+            System.currentTimeMillis()
+        );
+
+        return searchService.getRewriteContext(() -> System.currentTimeMillis(), resolvedIndices, null);
+    }
+
+    private class FullTextFunctionsRewritable implements Rewriteable<FullTextFunctionsRewritable> {
+
+        private final Map<FullTextFunction, QueryBuilder> queryBuilderMap;
+
+        FullTextFunctionsRewritable(Map<FullTextFunction, QueryBuilder> queryBuilderMap) {
+            this.queryBuilderMap = queryBuilderMap;
+        }
+
+        FullTextFunctionsRewritable(Set<FullTextFunction> functions) {
+            this.queryBuilderMap = new HashMap<>();
+
+            for (FullTextFunction func : functions) {
+                queryBuilderMap.put(func, func.asQuery(PlannerUtils.TRANSLATOR_HANDLER).asBuilder());
+            }
+        }
+
+        @Override
+        public FullTextFunctionsRewritable rewrite(QueryRewriteContext ctx) throws IOException {
+            Map<FullTextFunction, QueryBuilder> results = new HashMap<>();
+
+            boolean hasChanged = false;
+            for (FullTextFunction func : queryBuilderMap.keySet()) {
+                var initial = queryBuilderMap.get(func);
+                var rewritten = Rewriteable.rewrite(initial, ctx, false);
```

Comment:
No - because waiting for async actions will be done as part of this call:

https://github.com/elastic/elasticsearch/blob/e98d4705a814391e819665dcb3d3df0c012e36fb/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/QueryBuilderResolver.java#L74

Reply:
I missed that, thank you!